gtsam/basis/Basis.h

```diff
@@ -0,0 +1,373 @@
+
+    /// Constructor with interval [a,b]
+    EvaluationFunctor(size_t N, double x)
+        : weights_(DERIVED::CalculateWeights(N, x)) {}
+
+    /// Constructor with interval [a,b]
+    EvaluationFunctor(size_t N, double x, double a, double b)
+        : weights_(DERIVED::CalculateWeights(N, x, a, b)) {}
+
+    /// Regular 1D evaluation
+    double apply(const typename DERIVED::Parameters& f,
+                 OptionalJacobian<-1, -1> H = boost::none) const {
+      if (H) *H = weights_;
+      return (weights_ * f)(0);
+    }
+
+    /// c++ sugar
+    double operator()(const typename DERIVED::Parameters& f,
+                      OptionalJacobian<-1, -1> H = boost::none) const {
+      return apply(f, H);  // might call apply in derived
+    }
+
+    void print(const std::string& s = "") const {
+      std::cout << s << (s != "" ? " " : "") << weights_ << std::endl;
+    }
+  };
+
+  /**
+   * VectorEvaluationFunctor at a given x, applied to MatrixMN.
+   * This functor is used to evaluate a parameterized function at a given scalar
+   * value x. When given a specific M*N parameters, returns an M-vector the M
+   * corresponding functions at x, possibly with Jacobian wrpt the parameters.
+   */
+  template <int M>
+  class VectorEvaluationFunctor : protected EvaluationFunctor {
+   protected:
+    typedef Eigen::Matrix<double, M, 1> VectorM;
+    typedef Eigen::Matrix<double, M, -1> MatrixMN;
+    typedef Eigen::Matrix<double, /*MxMN*/ M, -1> Jacobian;
+    Jacobian H_;
+    void calculateJacobian() {
+      typedef Eigen::Matrix<double, M, M> MatrixM;
+      H_ = Eigen::kroneckerProduct(this->weights_, MatrixM::Identity());
+    }
+
+   public:
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+    // Used by FunctorizedFactor
+    using argument_type = Matrix;
+    using return_type = Vector;
+
+    /// Default Constructor
+    VectorEvaluationFunctor(size_t N, double x) : EvaluationFunctor(N, x) {
+      calculateJacobian();
+    }
+
+    /// Constructor, with interval [a,b]
+    VectorEvaluationFunctor(size_t N, double x, double a, double b)
+        : EvaluationFunctor(N, x, a, b) {
+      calculateJacobian();
+    }
+
+    /// M-dimensional evaluation
+    VectorM apply(const MatrixMN& F,
+                  OptionalJacobian</*M*N*/ -1, -1> H = boost::none) const {
+      if (H) *H = H_;
+      return F * this->weights_.transpose();
+    }
+
+    /// c++ sugar
+    VectorM operator()(const MatrixMN& F,
+                       OptionalJacobian</*M*N*/ -1, -1> H = boost::none) const {
+      return apply(F, H);
+    }
+  };
+
+  /**
+   * Given M*N Matrix of M-vectors at N Chebyshev points, predict component for
+   * given row i, with 0<i<M.
+   */
+  template <int M>
+  class ComponentEvaluationFunctor : public EvaluationFunctor {
+   protected:
+    typedef Eigen::Matrix<double, M, -1> MatrixMN;
+    typedef Eigen::Matrix<double, /*1xMN*/ 1, -1> Jacobian;
+    size_t rowIndex_;
+    Jacobian H_;
+    void calculateJacobian(size_t N) {
```

Comment:
Is this a manual Kronecker product?

Reply:
Nope. This is the evaluation of the jacobian which I did by hand and derived this.

Reply:
Add a comment what Jacobian is then? e.g.
```
/*
Calculate the `1*(M*N)` Jacobian of `apply` with respect to the M*N parameter matrix `P`.
We flatten assuming column-major order, e.g., if N=3 and M=2, we have 
	H=[w(0) 0  w(1) 0  w(2) 0] for rowIndex==0
	H=[0 w(0)  0 w(1)  0 w(2)] for rowIndex==1
i.e., one row of the Kronecker product of weights_ with the MxM identity matrix.
See also VectorEvaluationFunctor.
*/
```